---
title: "01_preprocessing_variant_data"
author: "Rebeca de Souza Brinco"
date: "April 13, 2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

## Data description
- _snp_list.csv_: List of SNPs related to resistance to biotic stress along with reference to their original articles and the pathogen/pest to which the SNPs are related. This database was created by manually curating the scientific literature on genome-wide association studies that aimed at identifying loci affecting disease/pest resistance. The search strategy used on Google Scholar was: _intitle:"soybean" AND "GWAS" AND "resistance"_,  _intitle:“soybean” AND “GWAS” AND “response”_, _intitle: “soybean” AND “GWAS” AND “ stress”_. Last update: April 2023.

## Data pre-processing and exploratory analyses

```{r}
library(BioNERO)
library(here)
library(tidyverse)
snplist <- read.csv(file= '/media/winterfell/rebeca/Project_2023/Project/Data/snp_list_only_some_biotic_stress_a2.csv', header=TRUE, sep=";") %>%
  select(!c(1:3)) %>%
  distinct() %>%
  mutate(SNP = str_trim(SNP)) %>%
  mutate(Stress = "biotic") 
```



Now, let's count the number of markers by category.

```{r}
table(snplist$MarkerType)
```

To make the downstream analyses easier, let's separate metaQTLs from SNPs.

```{r}
metaQTL <- snplist[snplist$MarkerType == "metaQTL", ]
snplist <- snplist[snplist$MarkerType != "metaQTL", ] # Remove metaQTLs
head(snplist)
nrow(snplist)
length(unique(snplist$SNP))
```

There are **`r nrow(snplist)` SNPs** on this database. Of these, `r length(unique(snplist$SNP))` are unique, which means that there are 2 SNPs shared by different studies.

Let's see how many SNPs we have per genome version.

```{r}
table(snplist$GenomeVersion)
```

Now, counting the SNPs per organism.

```{r}
sort(table(snplist$Organism), decreasing=TRUE)
```

## Getting genome positions for the SNPs

The .vcf file containing SNP positions for *Wm82.a2.v1* was downloaded from Soybase.

Load the file with the SNP positions.

```{r}
positions_a2 <- read.csv("/media/winterfell/rebeca/Project_2023/Project/Data/snp_positions_a2.v1.txt", 
                         header=T, sep = "\t")
names(positions_a2) <- c("chr", "pos", "id")
head(positions_a2)
```

Read in file with positions of SNPs starting with BARC*.


```{r}
barc_positions <- read.csv("/media/winterfell/rebeca/Project_2023/Project/Data/BARC_positions.txt", 
                           header=TRUE, sep="\t")
head(barc_positions)
```

Add data of more SNPs related to biotic and abiotic stresses

```{r}
snplist_biotic_and_abiotic <- read.csv("/media/winterfell/rebeca/Project_2023/Project/Data/snp_list_biotic_and_abiotic_stress_a2.v1.csv", header=T, sep = ",")
snplist_biotic_and_abiotic <- snplist_biotic_and_abiotic %>%
  select(Stress, Stress_info, Pathogen, DOI, SNP, MarkerType, GenomeVersion, Chr, Position) %>%
  mutate(Trait = snplist_biotic_and_abiotic$Stress_info) %>%
  select(!c(2, 6, 7)) %>%
  mutate(start = snplist_biotic_and_abiotic$Position) %>%
  mutate(end = snplist_biotic_and_abiotic$Position) %>%
  select(!6)

# rename NAs in pathogens to 0
snplist_biotic_and_abiotic[is.na(snplist_biotic_and_abiotic)] <- '0'


# Change the name of Pathogen and SNP column

colnames(snplist_biotic_and_abiotic)[2] <-'Organism'


snplist_biotic_and_abiotic
  
```


Combine `positions` and `barc_positions` in a single data frame. 

```{r}
pos1 <- positions_a2[, c(3,1,2, 2)]
names(pos1) <- c("SNP_ID", "Chr", "start", "end")
pos2 <- barc_positions[, c(1:4)]
names(pos2) <- names(pos1)

all_positions <- rbind(pos1, pos2)
head(all_positions)
```


Create new columns containing chromosome and position for the SNPs.

```{r}
snp_pos <- merge(snplist, all_positions, 
                 by.x="SNP", by.y="SNP_ID", all.x=TRUE)[, -c(5,6)]
head(snp_pos)
table(snp_pos$Chr)
```




Some SNPs do not have IDs. For these cases, SNP names are identified as `ChrX-pos`. Let's include genomic coordinates for these SNPs.

```{r}
sum(startsWith(snp_pos$SNP, "Chr"))
head(snp_pos[is.na(snp_pos$Chr), ])

# Add chromosome info
snp_pos$Chr[is.na(snp_pos$Chr)] <- sapply(strsplit(snp_pos$SNP[is.na(snp_pos$Chr)], "-"), `[`, 1)

# Add start and end positions
snp_pos$start[is.na(snp_pos$start)] <- sapply(
  strsplit(snp_pos$SNP[is.na(snp_pos$start)], "-"),
  `[`, 2
  ) 
snp_pos$end[is.na(snp_pos$end)] <- sapply(
  strsplit(snp_pos$SNP[is.na(snp_pos$end)], "-"), 
  tail, 1
  ) 

# Check if they have all been replaced
nrow(snp_pos[rowSums(is.na(snp_pos)) > 0, ])
```

# bind 'snp_pos' and 'snplist_biotic_and_abiotic'

```{r}
snp_pos_final <- rbind(snp_pos, snplist_biotic_and_abiotic)
```






To conclude, let's standardize chromosome names based on the `ChrX` nomenclature and save the SNP positions as GRanges and GRangesList objects.

```{r save_snp_granges}
library(GenomicRanges)
snp_ranges <- snp_pos_final %>%
  filter(!startsWith(Chr, "scaffold")) %>%
  mutate(Chr = str_replace(Chr, "Gm", "Chr")) %>%
  mutate(Chr = str_replace_all(Chr, 
                           c("Chr1$" = "Chr01",
                             "Chr2$" = "Chr02",
                             "Chr3$" = "Chr03", 
                             "Chr4$" = "Chr04", 
                             "Chr5$" = "Chr05", 
                             "Chr6$" = "Chr06", 
                             "Chr7$" = "Chr07",
                             "Chr8$" = "Chr08",
                             "Chr9$" = "Chr09",
                             "Chr10$" = "Chr10",
                             "Chr11$" = "Chr11",
                             "Chr12$" = "Chr12",
                             "Chr13$" = "Chr13",
                             "Chr14$" = "Chr14",
                             "Chr15$" = "Chr15",
                             "Chr16$" = "Chr16",
                             "Chr17$" = "Chr17",
                             "Chr18$" = "Chr18",
                             "Chr19$" = "Chr19",
                             "Chr20$" = "Chr20",
                             "^1$" = "Chr01",
                             "^2$" = "Chr02",
                             "^^3$" = "Chr03",
                             "^4$" = "Chr04",
                             "^5$" = "Chr05",
                             "^6$" = "Chr06",
                             "^7$" = "Chr07",
                             "^8$" = "Chr08",
                             "^9$" = "Chr09",
                             "^10$" = "Chr10",
                             "^11$" = "Chr11",
                             "^12$" = "Chr12",
                             "^13$" = "Chr13",
                             "^14$" = "Chr14",
                             "^15$" = "Chr15",
                             "^16$" = "Chr16",
                             "^17$" = "Chr17",
                             "^18$" = "Chr18",
                             "^19$" = "Chr19",
                             "^20$" = "Chr20"))
         ) %>%
  select(Chr, start, end, Stress, Organism, Trait, SNP, DOI)

# Remove NAs, i.e., remove SNPs without defined positions

snp_ranges2 <- na.omit(snp_ranges) 



```


# Let's divide the types of stress (biotic and abiotic) to save the SNP positions as GRanges and GRangesList objects.

```{r}

# Geral

# GRanges
snp_granges <- makeGRangesFromDataFrame(snp_ranges2)
snp_granges$Organism <- snp_ranges2$Organism
snp_granges$Trait <- snp_ranges2$Trait
snp_granges$SNP <- snp_ranges2$SNP
snp_granges$DOI <- snp_ranges2$DOI

# GRangesList
snp_grangeslist <- split(snp_granges, snp_granges$Trait)

# Saving data
save(snp_granges, snp_grangeslist, 
     file = "/media/winterfell/rebeca/Project_2023/Project/Products/result_files/snp_granges.rda", 
     compress = "xz")

# Biotic stress

snp_ranges_biotic <- snp_ranges2 %>%
  filter(startsWith(Stress, "biotic"))



# GRanges
snp_granges_biotic <- makeGRangesFromDataFrame(snp_ranges_biotic)
snp_granges_biotic$Organism <- snp_ranges_biotic$Organism
snp_granges_biotic$Trait <- snp_ranges_biotic$Trait
snp_granges_biotic$SNP <- snp_ranges_biotic$SNP
snp_granges_biotic$DOI <- snp_ranges_biotic$DOI

# GRangesList
snp_grangeslist_biotic <- split(snp_granges_biotic, snp_granges_biotic$Organism)

# Saving data
save(snp_granges_biotic, snp_grangeslist_biotic, 
     file = "/media/winterfell/rebeca/Project_2023/Project/Products/result_files/snp_granges_biotic.rda", 
     compress = "xz")




# Abiotic stress

snp_ranges_abiotic <- snp_ranges2 %>%
  filter(startsWith(Stress, "abiotic"))


# GRanges
snp_granges_abiotic <- makeGRangesFromDataFrame(snp_ranges_abiotic)
snp_granges_abiotic$Organism <- snp_ranges_abiotic$Organism
snp_granges_abiotic$Trait <- snp_ranges_abiotic$Trait
snp_granges_abiotic$SNP <- snp_ranges_abiotic$SNP
snp_granges_abiotic$DOI <- snp_ranges_abiotic$DOI
# GRangesList
snp_grangeslist_abiotic <- split(snp_granges_abiotic, snp_granges_abiotic$Trait)
# Saving data
save(snp_granges_abiotic, snp_grangeslist_abiotic, 
     file = "/media/winterfell/rebeca/Project_2023/Project/Products/result_files/snp_granges_abiotic.rda", 
     compress = "xz")
```



# To wrap it up, making a table of studies included in this study. This will be Table 1 in the manuscript.

```{r table1}
table1 <- as.data.frame(snp_granges)
table1 %>%
  count(DOI, Organism) %>%
  arrange(DOI, Organism)
```


## Session information

# This document was created under the following conditions:

```{r}
sessionInfo()
```

